### ISSUE TITLE
A JavaScript error occurred in the main process

### ISSUE THREAD

**Issue by ge0spawn** [aza547/wow-recorder#40]
i can't get it to record anything, 2s/3s/scrims

![image](https://user-images.githubusercontent.com/110961511/183822198-5f74a8eb-52e8-4ceb-a829-25869bdf1c3f.png)

**Comment by ge0spawn** [comment]
after the match is over, this error pops up

**Comment by aza547** [comment]
Thanks for the report. Can you share a screenshot of your settings, and the contents of the folder that is meant to contain the recordings? Interested in if any mp4 files are there.

**Comment by aza547** [comment]
To share some details, the function that is going wrong is looking for a video file to write a corresponding data file alongside with some details on the recording (who won etc). 

But it looks likely that there is no recording file created at all. Given you seem to be the only person who has hit this so far I suspect it might be related to your settings and PC setup (is the disk you are trying to save to writable? Do you have the correct permissions to do so?)

**Comment by aza547** [comment]
If your disk is really slow that could also be a bug that we don't wait long enough. Regardless we could do with some better error handling here.

**Comment by ge0spawn** [comment]
ah good point, let me change the location to an SSD first and see what happens

**Comment by ge0spawn** [comment]
the folder is empty, i've tried to put on 2 other different drives and even created it through the settings window when you select it.

![image](https://user-images.githubusercontent.com/110961511/184007014-e38a59d8-d075-4d61-a038-fe8efa7f0ab3.png)

**Comment by aza547** [comment]
Nope, doesn't need to be on the same drive. Settings look OK to me what you have. No luck with the SSD switch?

I wondered if it was a silly bug like spaces can't be included in the path, but I tested it myself and works fine for me with spaces. 

Are you familiar with using command line tools? I could give you some instructions on how to get more information here. Alternatively we could hop on a discord screen share if you could find the time - would like to get to the bottom of this!

**Comment by aza547** [comment]
If you can manage the command line tool stuff, here's the steps:

1. Right click the shortcut.
2. Select properties.
3. Select open file location. Windows explorer should open at the location. 
4. Type "powershell" in the bar with the folder location and hit enter. Powershell should open at the location. 
![image](https://user-images.githubusercontent.com/32389325/184020896-2e868cc0-7f2b-41d9-87c2-df5a57856e42.png)
6. Type `.\WarcraftRecorder.exe` and hit enter in the powershell window. Warcraft Recorder should open.
![image](https://user-images.githubusercontent.com/32389325/184020759-97afc718-7da2-4e23-ac9b-78711b7de3c0.png)

7. Test a skirmish.
8. Observe what is written in the powershell window, and paste it back here.  
9. Might look something like this:
![image](https://user-images.githubusercontent.com/32389325/184021060-1a0001d6-8ada-4b2b-af39-3bdf1d72ac64.png)

**Comment by aza547** [comment]
The other thing that would be interesting to see is the OBS logs. You can find these within the install location of the application, for me that is:
`C:\Users\alexa\AppData\Local\Programs\WarcraftRecorder\resources\app.asar.unpacked\dist\main\osn-data\node-obs\logs`

**Comment by ge0spawn** [comment]
ok, here we go 
![image](https://user-images.githubusercontent.com/110961511/184461270-a8e1331c-a058-4fde-9465-fcd021e410e2.png)

**Comment by ge0spawn** [comment]
and here is the log from a 3s match

[2022-08-12 20-42-56.txt](https://github.com/aza547/wow-recorder/files/9330478/2022-08-12.20-42-56.txt)

**Comment by aza547** [comment]
Hm I think this line is the issue:
```
[000:00:07:48.198.030.100][Error] [AMF] [H264]<Id: 3> <Plugin::AMD::EncoderH264::SetTargetBitrate> Failed to set to 0, error AMF_OUT_OF_RANGE (code 5)
```
The target bitrate sounds a bit like some settings we set manually here:
https://github.com/aza547/wow-recorder/blob/main/src/main/obsRecorder.ts#L71

I wonder if for some reason this doesn't work on your AMF encoder. That seems true:
https://github.com/obsproject/obs-amd-encoder/wiki/Configuration#target-bitrate-peak-bitrate

Looks like we are setting this to zero, and for most encoders that's fine and I guess is defaulting to lossless recording.

From my OBS logs:

```
[000:00:00:11.858.896.400][Info] [jim-nvenc: 'recording_h264'] settings:
[000:00:00:11.858.896.400][Info] 	rate_control: CBR
[000:00:00:11.858.896.400][Info] 	bitrate:      0
[000:00:00:11.858.896.400][Info] 	cqp:          20
[000:00:00:11.858.896.400][Info] 	keyint:       250
[000:00:00:11.858.896.400][Info] 	preset:       hq
[000:00:00:11.858.896.400][Info] 	profile:      high
[000:00:00:11.858.896.400][Info] 	width:        1920
[000:00:00:11.858.896.400][Info] 	height:       1080
[000:00:00:11.858.896.400][Info] 	2-pass:       false
[000:00:00:11.858.896.400][Info] 	b-frames:     2
[000:00:00:11.858.896.400][Info] 	lookahead:    false
[000:00:00:11.858.896.400][Info] 	psycho_aq:    true
```

Changing
`setSetting('Output', 'Recbitrate', 'lossless');` to `setSetting('Output', 'Recbitrate', 15000);` changes that 0 to 15000 in my logs. 

I'm going to build you a version of the app to see if I hardcode a number here if it will resolve your problems.

**Comment by aza547** [comment]
Can you take this build and re-test please?
https://www.dropbox.com/s/5z1l4aivo8qb5ef/WarcraftRecorder%20Setup%202.2.0.exe?dl=0

It has some specific changes in it to try make your encoder work, and some extra logging.

If you could repeat the powershell process above and share the results with me that would be awesome.

I have added some extra logging so that we can see exactly what your encoder is (the proper fix might be to manually set a non-zero bit-rate for your AMF encoder). 

Thanks!

**Comment by aza547** [comment]
From TMB:

```
Initializing OBS...
OBS initialized
Configuring OBS
Available encoder: ["none","obs_x264","amd_amf_h264"]
Selected encoder: amd_amf_h264
OBS Configured
```

**Comment by aza547** [comment]
[2022-08-14_23-32-39.txt](https://github.com/aza547/wow-recorder/files/9339179/2022-08-14_23-32-39.txt)

**Comment by aza547** [comment]
OK, fixes are in here:
https://www.dropbox.com/s/2egfnc43t58pd7o/WarcraftRecorder%20Setup%202.3.1-a1.exe?dl=0

**Comment by aza547** [comment]
Wonder if a better fix long term is to swap to VBR, with something like this?

```
     // Setting rate control to VBR
            obs.setSetting(EOBSSettingsCategories.Output, 'rate_control', 'VBR');
            obs.setSetting(EOBSSettingsCategories.Output, 'Recrate_control', 'VBR');
```

**Comment by aza547** [comment]
Fixed in 2.3.1. Closing this. Can come back to it if there is still problems. For now, AMF encoders have a hardcoded 50k bitrate. Might mean files are a bit bigger than they have to be for AMF encoders on small resolutions. That doesn't seem like the end of the world.
